Ge21 generalizing

dteague opened the pull request:
PR for changing the gem plotting code to be compatible with GE21 plotting

## Description
The gem-plotting code is hardcoded with many magic numbers that are specific to the GE11, so with changes in the tree structure to add gemtype and detectortype, the plotting code can be expanded to allow for GE21 plotting that is determined on the fly by this gemType variable.

The bulk of the code is removing instances of 24, 8, 3, and combinations of these to more generic variables, ie nVFATS, maxiEta, maxiPhi. These variables are derived from `chamber_maxiEtaiPhiPair` in  `./mapping/chamberInfo.py` and from `vfatsPerGemVariant`  in `gempython.tools.hw_constants` All of the mapping/chamberInfo.py variables have been turned into dictionaries that require the gemType to get the correct variables, ie

 ```chamber_vfatPos2iEta   =>    chamber_vfatPos2iEta["ge11"]```
etc.

Since a large code change is the actual canvas creation code, the largest change is to canvas making. The `makeAxBCanvas` has been removed in place of merging this functionality into the `saveSummaryCanvas` to make a new function called `getSummaryCanvas`.  This means the function now always returns a canvas created and there is a flag to decide to save the canvas or not. ie function def is now

```def getSummaryCanvas(dictSummary, dictSummaryPanPin2=None, name='Summary', trimPt=None, drawOpt="colz", gemType="ge11", write2Disk=False)
canv = make3x8Canvas(...)    =>    canv = getSummaryCanvas(..., gemType="ge11")
saveSummaryCanvas(...)    =>    getSummaryCanvas(..., gemType="ge11", write2Disk=True)
```

Also, because the `makeAxBCanvas` function also added histograms to an existing canvas, this functionality has be put to a helper function `addPlotToCanvas` which hopefully will make the code more readable as well. So the change becomes

```
canv = make3x8Canvas(initialContent=hist1, ...)
canv = make3x8Canvas(initialContent=hist2, canv=canv, ...)

=>

canv = getSummaryCanvas(hist1, ..., gemType="ge11")
canv = addPlotToCanvas(canv, hist2, gemType="ge11")
```

Last, many functions include now a gemType variable that needs to be supplied to give the right gemType. The default is "ge11", so the code remains backwards compatible for all cases where the gemType isn't actually supplied.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->
This change has mainly be motivated to be backwards compatible and to default to ge11 in all unknown cases. Since ge21 is needing plots specifically made for it (and not ones derived from ge11 algorithms), having code that can flexibly handle ge21 and in the future, me0 is a must.

For further information on this, look at issue #226 

## How Has This Been Tested?
For testing, the code was run over ge11 and ge21 files with the internal gemType flag flipped (ge21 files didn't have that tree in it). 

NOTE: currently, only the main run files (ie files in the top directory) are changed. Files in the `./macros` area have not been touched. Here is a list of what has been done.

- `anaDACScan.py`, `anaUltraScurve.py`, `anaSBitThresh.py`, `anaUltraLatency.py` - work for both
- `anaUltraThreshold.py`, `anaSBitMonitor.py` - work for ge11, could not fine ge21 file
- `anaSBitReadout.py`, `anaXDAQLatency.py` - could not find input file for either
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

jsturdy commented on anaSBitMonitor.py:
```diff
@@ -40,6 +42,16 @@
+    #### VERY TEMPORARY
+    from gempython.gemplotting.mapping.chamberInfo import gemTypeMapping
+    if 'gemType' not in inF.sbitDataTree.GetListOfBranches():
+        gemType = "ge21"
```
What do `NEED TO FIX` and `VERY TEMPORARY` mean in the context of this PR?
Also, here it appears that a default behaviour is going to be `GE2/1` which should not be the case (all changes should be completely transparent to the current operations)

dteague replied:
I put these comments because I was unsure the formatting of the gemType variable in the tree. I formatted it as a branch with the name "gemType" containing a number that mapped ge11 => 0, ge21 => 1. I didn't think this was how the final code would be done.

Is there a gemType branch implemented in the current tree structure, and if so, what is the format? I can change all these "TEMPORARY"s to that. 

Also, ge21 must have slipped by after some tests, will change

jsturdy replied:
OK, keep a `FIXME` text in each side of the comment block, so that it can easily be found when doing the eventual cleanup

dteague replied:
Done

jsturdy commented on anaSBitMonitor.py:
```diff
@@ -88,10 +102,10 @@
                         "h_vfatObservedVsVfatPulsed_{0}_{1}Hz".format(postScript,int(rate)),
                         "Summmary - Rate {0} Hz;VFAT {1};VFAT Observed".format(int(rate),strPulsedOrUnmasked),
-                        24,-0.5,23.5,24,-0.5,23.5)
+                        vfatsPerGemVariant[gemType],-0.5,vfatsPerGemVariant[gemType]-0.5, vfatsPerGemVariant[gemType],-0.5,vfatsPerGemVariant[gemType]-0.5)
```
probably, this lookup can be moved outside the loop, no?

dteague replied:
defined variable nVFATS and used that in this (and other places)

jsturdy commented on anaSBitMonitor.py:
```diff
@@ -138,7 +152,7 @@
                     dict_h_chanVsRatePulsed_ZRateObs[isValid][calEnable][vfat].GetXaxis().SetBinLabel(binX+1,formatSciNotation(str(rate)))
 
-                    # VFAT level
+                    # VFAT levpel
```
typo

dteague replied:
My errant emacs commands are showing :grin: Will Fix

jsturdy commented on anaSBitReadout.py:
```diff
@@ -241,9 +245,9 @@
-        ieta_h_strip[ieta] = r.TH1F("h_ieta{0}_strips_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), 384, 0., 384.)
-        ieta_h_ch[ieta] = r.TH1F("h_ieta{0}_chan_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), 384, 0., 384.)
+    for ieta in range(1, maxiEta+1):
+        ieta_h_strip[ieta] = r.TH1F("h_ieta{0}_strips_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), 128*maxiPhi, 0., 128.*maxiPhi)
```
Probably good to have a `CHANNELS_PER_VFAT` constant defined somewhere, to replace this repetition of the magic number `128` (where `128` refers to the number of VFAT channels...)

dteague replied:
I'm putting this variable in `mapping/ChamberInfo.py` for now, I'll implement that now. This can also be done for strips, ie a STRIPS_PER_IPHI variable if we want
Or is the max strips always going to be equal to the max channel number?

jsturdy replied:
I think you can safely assume that here `128` should always be `CHANNELS_PER_VFAT`, as the partitions are almost surely defined by the locations of the VFATs

mexanick replied:
With GE2/1 the eta mapping would be different, as each VFAT will control 64 strips in one eta partition and 64 strips in another, adjacent eta partition in a way, that a trigger pad will consist of 2 strips with the same phi and different eta coordinates. Thus said, the number of channels per VFAT will stay of course the same.

dteague replied:
To make sure I understand how we want to implement this: because we are going to have an equal number of strips to vfats, we should just replace all instances of 128 with CHANNELS_PER_VFAT since the distinction won't matter much, eg:
```
ieta_h_strip[ieta] = r.TH1F("h_ieta{0}_strips_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), 128*maxiPhi, 0., 128.*maxiPhi)
...
ieta_h_strip[eta].Fill((phi-1)*128+strip[0])
```

If this is correct, I'll go ahead and make that change

dteague replied:
Followed instructions from meeting

jsturdy commented on anaSBitReadout.py:
```diff
@@ -366,57 +371,46 @@
     # Summaries Canvas
     #
     # make3x8Canvas
```
you've changed the semantics, i'd change/remove the comment

dteague replied:
done

jsturdy commented on anaSBitReadout.py:
```diff
@@ -425,7 +419,7 @@
 
     ietaDir.cd()
-    for ieta in range(1, 9):
+    for ieta in range(1, maxiEta):
```
this won't reproduce the same behaviour `maxiEta != 9`
possibly more readable would be `range(maxiEta):` and in the loop the value is used as `ieta+1`

dteague replied:
fixed

jsturdy commented on anaXDAQLatency.py:
```diff
@@ -45,6 +45,19 @@
+    else:
+        gemType = gemTypeMapping[rp.tree2array(tree=inFile.latTree, branches =[ 'gemType' ] )[0][0]]
+    print gemType
+    ##### END
```
again, what does this comment mean in the context of this PR?

dteague replied:
See above comment on reading in gemType

jsturdy commented on anaXDAQLatency.py:
```diff
@@ -74,7 +87,7 @@
     from gempython.utils.nesteddict import nesteddict as ndict
     baseDir = ndict() # baseDir[slot][oh] -> string
-    vfatDirs = ["VFAT-%d"%x for x in range(24)]
+    vfatDirs = ["VFAT-%d"%x for x in range(vfatsPerGemVariant[gemType])]
```
two things:
1) as above, it's probably best to set this `vfatsPerGemVariant` lookup somewhere outside of all the loops, so it's only done once.
1) i would *really* love to get rid of these poorly named `TDirectories` (using `-` in the name breaks a lot of interactive stuff (but this change can't be made here as it would break backwards compatibility)

dteague replied:
I made a dedicated `nVFATS` variable for most functions and have them use that instead of the dictionary.

jsturdy commented on mapping/chamberInfo.py:
```diff
@@ -34,8 +34,12 @@
+##### NEED TO FIX
+#### VERY TEMPORARY
+gemTypeMapping = {0:"ge11", 1:"ge21"}
+##### END    
```
again

dteague replied:
This is the adhoc mapping I made for the gemType branch: see above comment are the branch's format

jsturdy commented on utils/latAlgos.py:
```diff
@@ -71,17 +71,32 @@ def anaUltraLatency(infilename, debug=False, latSigMaskRange=None, latSigRange=N
+    else:
+        gemType = gemTypeMapping[rp.tree2array(tree=inFile.latTree, branches =[ 'gemType' ] )[0][0]]
+    print gemType
+    ##### END
```
again

dteague replied:
as above

jsturdy commented on utils/latAlgos.py:
```diff
@@ -129,13 +144,15 @@ def anaUltraLatency(infilename, debug=False, latSigMaskRange=None, latSigRange=N
     hHitsVsLat_AllVFATs.SetTitle("Sum over all VFATs")
-    for vfat in range(1,24):
+    #### QUESTION 
+    # why is this only [1-24)
```
because it is cloned in line 151 from the VFAT0 histogram

dteague replied:
:heavy_check_mark:

jsturdy commented on utils/scurveAlgos.py:
```diff
@@ -137,40 +137,48 @@ def anaUltraScurve(args, scurveFilename, calFile=None, GEBtype="short", outputDi
+    from gempython.gemplotting.mapping.chamberInfo import gemTypeMapping
+    gemType = gemTypeMapping[rp.tree2array(scurveTree, branches =[ 'gemType' ] )[0][0]]
+    print gemType
+    ##### END
```
again

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -128,17 +116,41 @@ def anaUltraThreshold(args,thrFilename,GEBtype="short",outputDir=None,fileScurve
+    else:
+        gemType = gemTypeMapping[rp.tree2array(tree=thrTree, branches =[ 'gemType' ] )[0][0]]
+    print gemType
+    ##### END
```
again

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -553,15 +562,22 @@ def calibrateThrDAC(args):
+    # ##### NEED TO FIX
+    # #### VERY TEMPORARY
+    gemType = chamberName[:chamberName.find("-")].lower()
+    # ##### END
```
again

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -1110,21 +1126,24 @@ def calibrateThrDAC(args):
-    canvScurveSigmaVsThrDac_BoxPlot_Summary = make3x8Canvas("canvScurveSigmaVsThrDac_BoxPlot_Summary",dict_ScurveSigmaVsThrDac_BoxPlot,"candle1")
+    canvScurveMeanByThrDac_Summary = getSummaryCanvas(dict_mGraphScurveMean, name="canvScurveMeanByThrDac_Summary", drawOpt="APE1", gemType=gemType)
+    canvScurveSigmaByThrDac_Summary = getSummaryCanvas(dict_mGraphScurveSigma, name="canvScurveSigmaByThrDac_Summary", drawOpt="APE1", gemType=gemType)
+    ### FIX THIS
```
fix what?

dteague replied:
AH, this was a comment before i added the `addPlotToCanvas` function. Before, the `make3x8Canvas` function was called multiple times on the same canvas to fill all the histograms, so my note was make a funciton to reproduce this! Removed the old comment

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -1260,6 +1280,16 @@ def getDetName(entry):
+    else:
+        gemType = gemTypeMapping[rp.tree2array(tree=rateTree, branches =[ 'gemType' ] )[0][0]]
+    print gemType
+    ##### END
```
again

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -1288,12 +1318,13 @@ def getDetName(entry):
         for entry in crateMap:
             ohKey = (entry['shelf'],entry['slot'],entry['link'])
-            for vfat in range(0,25): #24th VFAT represents "overall case"
+            for vfat in range(0,vfatsPerGemVariant[gemType]+1): #24th VFAT represents "overall case"
```
unfortunately, there are two conventions for the "all VFATs summary" plot, one is to put it in index -1, one is to put it in index 24
this should be cleaned up everywhere, but this is outside this PR

jsturdy commented on utils/threshAlgos.py:
```diff
@@ -1303,7 +1334,7 @@ def getDetName(entry):
                 dict_Rate1DVsDACNameX[dacName][ohKey][vfat].SetMarkerSize(0.8)
                 dict_Rate1DVsDACNameX[dacName][ohKey][vfat].SetLineWidth(2)
-
+                print dacName, ohKey, vfat
```
debugging prints should be removed

jsturdy commented:
when you go to address comments, you should `rebase` onto the latest `cms-gem-daq-project/develop`

dteague commented:
- I changed most all of the print statements to python 3 format
- I changed most all of the % string formatting to the preferred `.format()` function
- I made a variable for many function for max vfats and max channels (`nVFATS` and `maxChans`). - Of course the naming can be changed easily, but it removes multiple look ups of the vfat dictionary and remove some 128 magic numbers from the code 

I do have some questions as shown by the comments to code blocks, and after those are answered, I can fix and test on files again to make sure everything is working correctly

jsturdy commented:
OK, it would probably be good to sit together and go over the `rebase` workflow, as your previous attempt succeeded (but in a sub-optimal way, as you created a new merge and now have a bunch of duplicate commits)

dteague commented:
I looked into this and I removed the extraneous commits and rebased them on the newest changes. The commit log shows only the 6 commits on top of the develop branch, so all seems to have worked, but this might need to be looked over since my git skills are still pretty poor :stuck_out_tongue_closed_eyes:

jsturdy commented on anaSBitReadout.py:
```diff
@@ -312,16 +322,16 @@
             # filling ieta 1Dhistos
             ieta_h_strip[eta].Fill((phi-1)*128+strip[0])
```
Likewise here, `128` is `CHANNELS_PER_VFAT` (also L331, L332, L365, L367...)

dteague replied:
done

jsturdy commented on anaSBitReadout.py:
```diff
@@ -241,9 +250,9 @@
+    for ieta in range(1, maxiEta+1):
+        ieta_h_strip[ieta] = r.TH1F("h_ieta{0}_strips_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), 128*maxiPhi, 0., 128.*maxiPhi)
+        ieta_h_ch[ieta] = r.TH1F("h_ieta{0}_chan_vs_hit".format(ieta), "i#eta = {0} | i#phi (1,2,3)".format(ieta), maxChans*maxiPhi, 0., maxChans*maxiPhi)
         ieta_h_sbitSize[ieta] = r.TH1F("h_ieta{0}_sbitSize_vs_hit".format(ieta), "i#eta = {0} SBIT Size".format(ieta), 8, 0., 8.)
```
~~Here *possibly* `8`-> `NUM_IETA_SECTORS`~~
Actually, I think it's the s-bit partition, which is not tied to the `ieta` sectors

jsturdy commented on anaSBitReadout.py:
```diff
@@ -255,11 +264,12 @@
+    
     from gempython.utils.gemlogger import printRed, printYellow
     print("Analyzing Raw Data\nThis may take some time please be patient")
     h_clusterMulti = r.TH1F("h_clusterMulti".format(vfat), "", 9,-0.5,8.5)
```
Here *possibly* `9`-> `NUM_IETA_SECTORS+1`, probably best to check the semantics when the `Fill` is done

dteague replied:
Hard to tell, but I don't think. This variable is based on a loop that runs over the clusterNames which are derived from the branches in file, ie 
https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/c987b20144e9e54619277a40c1d57772e5648474/anaSBitReadout.py#L282-L295

and 

https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/c987b20144e9e54619277a40c1d57772e5648474/anaSBitReadout.py#L190-L193

jsturdy commented on fitting/fitScanData.py:
```diff
@@ -313,7 +314,7 @@ def fit(self, debug=False):
                     if debug:
                         if self.isVFAT3:
-                            print "| %i | %i | %i | %i | %f | %f | %f | %f | %f | %f | %f | %f | %f |"%(
+                            print("| {0} | {1} | {2} | {3} | {4} | {5} | {6} | {7} | {8} | {9} | {10} | {11} | {12} |".format(
```
It wasn't my intention with my comment to have you do **all** the conversions, (just the ones that were in code you added/modified), but it's a good thing to have done anyway :-)
However, here does the formatting stay constant?
I suppose it can guess at the type, just need to make sure there isn't any implicit type conversion happening during the formatting...

dteague replied:
I ended up doing a replace for the files I edited, and changed up the numbers when need be! I can add types to these just to be on the safe side

dteague commented on utils/scurveAlgos.py:
```diff
@@ -1075,6 +1095,7 @@ def fill2DScurveSummaryPlots(scurveTree, vfatHistos, vfatChanLUT, vfatHistosPanP
             if (stripPinOrChan < 64):
+                #### Off by one error...?
                 vfatHistos[event.vfatN].SetBinContent(63-(stripPinOrChan+1),chargeBin,event.Nhits)
                 vfatHistos[event.vfatN].SetBinError(63-(stripPinOrChan+1),chargeBin,sqrt(event.Nhits))
```
This seems like an off by one error, namely the (64-1) and (stripPinOrChan+1) seem to be in opposition, eg 
```stripPinOrChan = 0   =>   bin = 62 ```
```stripPinOrChan = 63   =>   bin = -1```
Want to confirm before I try and change this

jsturdy replied:
I don't know...
* What is the valid range of values for `stripPinOrChan`?
* What do the current histograms look like (especially if you show the over/underflow stats)
* All things being equal, it looks to me like the `+1` should be outside the parentheses, and maybe also `vfatHistos` should be `vfatHistosPanPin2`...

dteague replied:
I did some tests, and having the +1 outside was correct. stripPinOrChan in {0,...,127} and the code works by if done by PanPin, the channels are split between the two histograms, `vfatHistos` and `vfatHistosPanPin2`, each having a domain of [0,63].

All that being said, I made the fix!

dteague commented:
Ran tests and made minor changes. Tests are the same as before, namely:

- anaDACScan.py, anaUltraScurve.py, anaSBitThresh.py, anaUltraLatency.py - work for both
- anaUltraThreshold.py, anaSBitMonitor.py - work for ge11, could not fine ge21 file
- anaSBitReadout.py, anaXDAQLatency.py - could not find input file for either

I did notice that `anaSBitMonitor` gave strange results for GE11, mainly that all vfats pointed to vfat 7 and were misID'd the same number of times for each vfat. Considering not much was changed with this file, I'm inclined to believe this is a problem with the file I'm running over, not the changes made

All of the gemType reading in is surrounded by comments saying `##### FIXME` so when a proper gemType read in is decided, that can be changed quickly. For all of these tests, I just changed the default value to "ge21" and recompiled.

If anyone has files for run over the files I haven't checked or wants other files, namely in the `macros` area changed, I can do that. Otherwise, I hopefully have fixed for all of the above comments.

jsturdy commented:
> Considering not much was changed with this file, I'm inclined to believe this is a problem with the file I'm running over, not the changes made

You can verify by using the unmodified code to run the same analysis on the same file (since it's GE1/1)

dteague commented:
I ran the code in the develop branch, and it seems like this code isn't used much because it doesn't work out of the box! (`args.filename` should be `args.infilename`. It is fixed in this PR). But the develop branch version gives the exact same results, so this is just a strange fluke of the file I think.

jsturdy commented:
> I ran the code in the develop branch, and it seems like this code isn't used much because it doesn't work out of the box! (`args.filename` should be `args.infilename`. It is fixed in this PR). But the develop branch version gives the exact same results, so this is just a strange fluke of the file I think.

@AndrewLevin there was some issue in the past that was related to this `filename` vs `infilename` if I recall correctly, was this resolved?  is the change that @dteague is making here counter to that discussion?
(@dteague, you can probably search the issues for that string to find what the discussion was about)

dteague commented:
Pinging for status
Also with #256 , it might be good to fix this concurrently with this PR else GE21 plots can't be graphed (will look into fixes and possibly make another PR)

mexanick commented:
@dteague the issue #256 requires a separate PR. Would be great if you can provide it

AndrewLevin commented:
> > I ran the code in the develop branch, and it seems like this code isn't used much because it doesn't work out of the box! (`args.filename` should be `args.infilename`. It is fixed in this PR). But the develop branch version gives the exact same results, so this is just a strange fluke of the file I think.
> 
> @AndrewLevin there was some issue in the past that was related to this `filename` vs `infilename` if I recall correctly, was this resolved? is the change that @dteague is making here counter to that discussion?

I think maybe you are referring to this change

https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/239/files#diff-058f061e69c295595d79241cf963cf20L105-R106

but that has nothing to do with this.

> (@dteague, you can probably search the issues for that string to find what the discussion was about)

jsturdy commented:
> > > I ran the code in the develop branch, and it seems like this code isn't used much because it doesn't work out of the box! (`args.filename` should be `args.infilename`. It is fixed in this PR). But the develop branch version gives the exact same results, so this is just a strange fluke of the file I think.
> > 
> > 
> > @AndrewLevin there was some issue in the past that was related to this `filename` vs `infilename` if I recall correctly, was this resolved? is the change that @dteague is making here counter to that discussion?
> 
> I think maybe you are referring to this change
> 
> https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/239/files#diff-058f061e69c295595d79241cf963cf20L105-R106
> 
> but that has nothing to do with this.
> 
> > (@dteague, you can probably search the issues for that string to find what the discussion was about)

OK, thanks @AndrewLevin, I just wanted to make sure that this change was properly motivated and not rehashing something historical.

jsturdy commented:
@dteague, sorry, can you `rebase` this one final time?
just want to make sure it would be a good merge as you've touched lots of files

dteague commented:
Just took care of that. Hopefully I rebased correctly and all is up to snuff!

lpetre-ulb commented:
While testing the new GE2/1 setup at FIT, Stephen and I encountered an issue with the usage of this PR within the [GE2/1 compatibility PR in `vfatqc-python-tools`](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265).

Indeed, in `getVFAT3CalInfo` the `gemType` argument is not set in `testConnectivity.py` nor in any other script; this is an issue for GE2/1 operation. While this seems an issue related to `vfatqc-python-tools` PR, the design in `dbutils.py` generalization could be improved to prevent this kind of problem.

In fact, why is the GEM type even required to be given to the DB util functions? I mean, one should be able to request the data for an arbitrary number of VFATs; 24 for GE1/1, 12 for GE2/1 or n for a manual lookup.

I see that `gemType` is used only here:
https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/6612de6cc575eef2b588bc4678c559c80a13d6a5/utils/dbutils.py#L176

Couldn't `vfatsPerGemVariant[gemType]` be advantageously replaced with `len(vfatList)`? This would make the DB util functions fully generic without any requirement on the GEM type. As a side effect, the PR in `vfatqc-python-scripts` would work without modification.

dteague commented:
>Couldn't vfatsPerGemVariant[gemType] be advantageously replaced with len(vfatList)? This would make the DB util functions fully generic without any requirement on the GEM type. As a side effect, the PR in vfatqc-python-scripts would work without modification.

Very fair, @lpetre-ulb, that is a much better implementation that I didn't see! Let me fix that quickly

lpetre-ulb commented:
Thank you for the quick fix @dteague!

jsturdy commented:
~~@mexanick, do we want to hold this while @sbutalla tests at FIT, or just merge it as is and do bugfixes later?~~
oops! too slow of a refresh! (thanks @lpetre-ulb!)

lpetre-ulb commented:
@sbutalla and I discovered what seems like a bug yesterday in the DAC scans plots for GE2/1 (at least). The title is a fifth order polynomial and the axis do not have any legend. Stephen, can you open an issue? I don't have access to the plots.
